File: src/Julia/Magnet-particules-in-boxes.jl


using Test
using Printf


"""
通项公式为 1/k/(n+1)^(2k)  
对此公式交叉遍历 k 向量和 n 向量，求和
"""
function doubles(maxk::Int, maxn::Int)::Float64
    v(k, n) = 1 / k / (big(n) + 1)^(2k) # 不转换Float64或BigInt会数值越界
    [v(k, n) for k ∈ 1:maxk for n ∈ 1:maxn] |> sum
end

doubles(10, 1000)


function ifinrange(maxk::Int, maxn::Int, expect::Float64; tolerence=1e-6)::Bool
    # println("Expect $(round(expect, digits = 12))")
    @printf("Expect %.12f\n", expect)

    actual = doubles(maxk, maxn)
    @printf("Actual %.12f\n", actual)

    abs(actual - expect) ≤ tolerence
end



@testset "fixed tests" begin
    @test ifinrange(1, 10, 0.5580321939764581) == true
    @test ifinrange(10, 1000, 0.6921486500921933) == true
end

File: src/R/Magnet-particules-in-boxes.R
library(tidyverse)
library(data.table)
library(magrittr)


doubles <- function(maxk, maxn) {
  iterator <- expand.grid(k = 1:maxk, n = 1:maxn) %>% as.data.table()
  iterator[, v := 1 / (k * (n + 1)^(2 * k))][, v] %>% sum()
}

doubles(1, 10)



library(testthat)
dotest <- function(maxk, maxn, expected) {
  actual <- doubles(maxk, maxn)
  expect_equal(actual, expected, tol = 1e-6, scale = 1, info = "")
}
test_that("test doubles", {
  dotest(1, 10, 0.5580321939764581)
  dotest(10, 1000, 0.6921486500921933)
  dotest(10, 5000, 0.6929472124031200)
})
